plugins/removeUnusedNS.js

```diff
@@ -1,82 +1,59 @@
 'use strict';
 
-const { traverse } = require('../lib/xast.js');
-const { parseName } = require('../lib/svgo/tools.js');
-
+exports.type = 'visitor';
 exports.name = 'removeUnusedNS';
-
-exports.type = 'full';
-
 exports.active = true;
-
 exports.description = 'removes unused namespaces declaration';
 
 /**
  * Remove unused namespaces declaration.
  *
- * @param {Object} item current iteration item
- * @return {Boolean} if false, item will be filtered out
- *
  * @author Kir Belevich
+ *
+ * @type {import('../lib/types').Plugin<void>}
  */
-exports.fn = function (root) {
-  let svgElem;
-  const xmlnsCollection = [];
-
+exports.fn = () => {
   /**
-   * Remove namespace from collection.
-   *
-   * @param {String} ns namescape name
+   * @type {Set<string>}
    */
-  function removeNSfromCollection(ns) {
-    const pos = xmlnsCollection.indexOf(ns);
-
-    // if found - remove ns from the namespaces collection
-    if (pos > -1) {
-      xmlnsCollection.splice(pos, 1);
-    }
-  }
-
-  traverse(root, (node) => {
-    if (node.type === 'element') {
-      if (node.name === 'svg') {
-        for (const name of Object.keys(node.attributes)) {
-          const { prefix, local } = parseName(name);
-          // collect namespaces
-          if (prefix === 'xmlns' && local) {
-            xmlnsCollection.push(local);
+  const definedNamespaces = new Set();
+  return {
+    element: {
+      enter: (node, parentNode) => {
+        // collect namespaces from svg element
+        if (node.name === 'svg' && parentNode.type === 'root') {
+          for (const name of Object.keys(node.attributes)) {
+            if (name.startsWith('xmlns:')) {
+              const local = name.slice('xmlns:'.length);
+              definedNamespaces.add(local);
+            }
           }
         }
-
-        // if svg element has ns-attr
-        if (xmlnsCollection.length) {
-          // save svg element
-          svgElem = node;
-        }
-      }
-
-      if (xmlnsCollection.length) {
-        const { prefix } = parseName(node.name);
-        // check node for the ns-attrs
-        if (prefix) {
-          removeNSfromCollection(prefix);
+        if (definedNamespaces.size !== 0) {
+          // remove element namespace from collected list
+          if (node.name.includes(':')) {
+            const [ns] = node.name.split(':');
+            if (definedNamespaces.has(ns)) {
+              definedNamespaces.delete(ns);
+            }
+          }
+          // check each attr for the ns-attrs
+          for (const name of Object.keys(node.attributes)) {
+            if (name.includes(':')) {
+              const [ns] = name.split(':');
+              definedNamespaces.delete(ns);
+            }
+          }
```

Comment:
The code looks good, but it feels like it doesn't have enough explanation that allows other developers to understand it faster. For example, we remove some elements from `definedNamespaces` but there is a lack of explanation why we do that. Would be better to have reference to some specs or describe the goal: Examples:

 ```js
          // Collect all namespaces from the svg element
          // (such as `xmlns:xlink="http://www.w3.org/1999/xlink"`) 
          if (node.name === 'svg' && parentNode.type === 'root') {

          // We must preserve a namespace if there are nested elements that refer
          // to the same namespace in their names
          if (node.name.includes(':')) {
          // ...

          // Dropping a namespace from the root element might break the rendering of elements
          // that use this namespace in attribute names
          for (const name of Object.keys(node.attributes)) {
          // ...
```

Reply:
Tweaked comments